File: linkedlist/add_two_numbers/add_two_numbers.go
package addtwonumbers

type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	dummy := &ListNode{}
	tail := dummy
	carry := 0
	for !done(l1, l2, carry) {
		digit, nextCarry := addDigits(val(l1), val(l2), carry)
		tail = appendDigit(tail, digit)
		l1, l2, carry = advance(l1), advance(l2), nextCarry
	}
	return dummy.Next
}

func done(l1, l2 *ListNode, carry int) bool {
	return l1 == nil && l2 == nil && carry == 0
}

func val(n *ListNode) int {
	if n == nil {
		return 0
	}
	return n.Val
}

func advance(n *ListNode) *ListNode {
	if n == nil {
		return nil
	}
	return n.Next
}

func addDigits(a, b, carry int) (int, int) {
	sum := a + b + carry
	return sum % 10, sum / 10
}

func appendDigit(tail *ListNode, digit int) *ListNode {
	tail.Next = &ListNode{Val: digit}
	return tail.Next
}

File: linkedlist/add_two_numbers/add_two_numbers.ts
export class ListNode {
    val: number
    next: ListNode | null
    constructor(val?: number, next?: ListNode | null) {
        this.val = (val===undefined ? 0 : val)
        this.next = (next===undefined ? null : next)
    }
}
export function addTwoNumbers(l1: ListNode | null, l2: ListNode | null): ListNode | null {
    let dummy = new ListNode(0)
    let tail = dummy
    let carry = 0

    while (!done(l1, l2, carry)) {
        let [digit, nextCarry] = addDigits(getVal(l1), getVal(l2), carry)
        tail = appendDigit(tail, digit)
        carry = nextCarry
        l1 = advance(l1)
        l2 = advance(l2)
    }
    return dummy.next
};

function done(l1: ListNode | null, l2: ListNode | null, carry: number): boolean {
    return l1 === null && l2 === null && carry === 0
}

function getVal(node: ListNode | null): number {
    return node?.val ?? 0
}

function advance(node: ListNode | null): ListNode | null {
    return node?.next ?? null
}

function addDigits(a: number, b: number, carry: number): [number, number] {
    let sum = a + b + carry
    return [sum % 10, Math.floor(sum / 10)]
}

function appendDigit(tail: ListNode, digit: number): ListNode {
    tail.next = new ListNode(digit)
    return tail.next
}